Add support for generalizing and vectorizing linalg_ext.pack op.

hanhanW opened the pull request:
To vectorize the pack op, we generalize the pack op to tensor.pad + linalg.generic ops and kick in upstream generic vectorizer.

In the PR, we require outer dimensions of results to be all 1s. The restriction can be relaxed, but that requires us to add special vectorization pattern for expand_shape. The other way to support the cases is to tile outer dims, make them be all 1s and kick in generalization and generic vectorizer.

hanhanW commented:
@chelini please take a look, thanks!

nicolasvasilache reviewed (changes requested):
can we just tile a subset of the loops by 1 and rewrite as a linalg.generic and let the generic vectorizer kick in ?

nicolasvasilache commented:
> 1. The padding semantic won't go away even if we tile the loops by 1

We typically rewrite this as a fill + copy.

> 2. Even there are no padding semantics, rewrite it to a single generic op would have non-projected permutation indexing_map. The indexing maps would be something like (d0, d1, d2, d3) -> (d0 * s0 + d1, d2 * s1 + d3).

That is the whole point of tiling the tiled loops by 1 with a rank-reducing slice: `d0` and `d2` become `0` and the copy should simply be `(d0, d1) -> (d0, d1).`

Fill + copy are easy to vectorize.

hanhanW commented:
> > 1. The padding semantic won't go away even if we tile the loops by 1
> 
> We typically rewrite this as a fill + copy.
> 
> > 2. Even there are no padding semantics, rewrite it to a single generic op would have non-projected permutation indexing_map. The indexing maps would be something like (d0, d1, d2, d3) -> (d0 * s0 + d1, d2 * s1 + d3).
> 
> That is the whole point of tiling the tiled loops by 1 with a rank-reducing slice: `d0` and `d2` become `0` and the copy should simply be `(d0, d1) -> (d1, d0).`
> 
> Fill + copy/transpose are easy to vectorize.

sorry, I was wrong on indexing maps. That's probably the case for unpack. For pack op, we can rewrite it to `tensor.empty + linalg.fill + tensor.expand_shape + linalg.generic (transpose)`, and kick in the generic vectorize. One question is that I'm not sure if the vectorization of `tensor.expand_shape` is implemented or not. But that would be very straight forward to me  it's should just be replaced with `vector.shape_cast`.

thanks for the details here, and I'll take a stab at it!

hanhanW commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/Vectorization.cpp:
```diff
@@ -0,0 +1,179 @@
+      int64_t size = inputType.getDimSize(dim);
+      if (size % tileSize.getValue()) {
+        readInBounds[dim] = false;
+        readSizes.push_back(llvm::divideCeil(size, tileSize.getValue()) *
```
I was not aware of it, thanks! I'd keep using LLVM version because most of codes in IREE are using LLVM version. It's for consistency.

hanhanW commented:
I followed the suggestions from @nicolasvasilache please take a look, thanks!

chelini commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/Vectorization.cpp:
```diff
@@ -0,0 +1,255 @@
+    Value shapeCast =
+        rewriter.create<vector::ShapeCastOp>(loc, destVecType, read);
+
+    // TBD: it will be tensor::EmptyOp after an integration.
```
still needed?

chelini commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/Vectorization.cpp:
```diff
@@ -0,0 +1,255 @@
+  }
+};
+
+struct ExpandShapeVectorizationPattern
```
Can you please add some doc on this pattern?

hanhanW replied:
I added the comment, but it's not pushed... sorry about that. I'll push it to the branch now.

hanhanW commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/Vectorization.cpp:
```diff
@@ -0,0 +1,255 @@
+  }
+  auto resultType =
+      RankedTensorType::get(paddedShape, inputType.getElementType());
+  return tensor::createPadScalarOp(resultType, input, packOp.getPaddingValue(),
```
cool stuff, I was not aware of it. done!

MaheshRavishankar reviewed (changes requested):
I think we should restrict the vectorization to when the outer dimension of the result are all 1s? I am wondering if that would avoid the `shape_cast`?

MaheshRavishankar commented:
> I think we should restrict the vectorization to when the outer dimension of the result are all 1s? I am wondering if that would avoid the `shape_cast`?

Or to restate it, the `shape_cast` only introduces unit dimensions. I think that is the intent of the shape cast.

hanhanW commented:
If we tile outer dimensions of the result to 1s, the shape_cast will only introduce unit dimensions.

The doc states that the shape_cast allows introducing non-unit dimensions, but they are expected to be folded away before lowering them to LLVM or other dialects? Is it the reason about restricting outer dims to be all 1s? Because they  might be folded away by some rank-reduced ops?

```
    If expanding rank (n < k), source dimensions must factor into a
    contiguous sequence of destination dimension sizes.
    Each source dim is expanded (or contiguous sequence of source dims combined)
    in source dimension list order (i.e. 0 <= i < n), to produce a contiguous
    sequence of result dims (or a single result dim), in result dimension list
    order (i.e. 0 <= j < k). The product of all source dimension sizes and all
    result dimension sizes must match.

    It is currently assumed that this operation does not require moving data,
    and that it will be folded away before lowering vector operations.

    There is an exception to the folding expectation when targeting
    llvm.intr.matrix operations. We need a type conversion back and forth from a
    2-D MLIR vector to a 1-D flattened LLVM vector.shape_cast lowering to LLVM
    is supported in that particular case, for now.
```

MaheshRavishankar commented:
> If we tile outer dimensions of the result to 1s, the shape_cast will only introduce unit dimensions.
> 
> The doc states that the shape_cast allows introducing non-unit dimensions, but they are expected to be folded away before lowering them to LLVM or other dialects? Is it the reason about restricting outer dims to be all 1s? Because they might be folded away by some rank-reduced ops?
> 
> ```
>     If expanding rank (n < k), source dimensions must factor into a
>     contiguous sequence of destination dimension sizes.
>     Each source dim is expanded (or contiguous sequence of source dims combined)
>     in source dimension list order (i.e. 0 <= i < n), to produce a contiguous
>     sequence of result dims (or a single result dim), in result dimension list
>     order (i.e. 0 <= j < k). The product of all source dimension sizes and all
>     result dimension sizes must match.
> 
>     It is currently assumed that this operation does not require moving data,
>     and that it will be folded away before lowering vector operations.
> 
>     There is an exception to the folding expectation when targeting
>     llvm.intr.matrix operations. We need a type conversion back and forth from a
>     2-D MLIR vector to a 1-D flattened LLVM vector.shape_cast lowering to LLVM
>     is supported in that particular case, for now.
> ```

THe way I figure we will lower these is to tile the "outer tile dimensions" by one (after distribution) and just load store the individual tiles.... so in terms of prefered lowering path, we should vectorize after all the outer dimensions become 1. The same should happen for all tiled matmul ops (that is how mmt4d is handled today), and kind of what conv does today too... So lets start with that requirement. We can relax if we need to....

hanhanW commented:
I see your point now, thanks! I'll add the restriction and simplify the code.

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/Vectorization.cpp:
```diff
@@ -35,74 +37,128 @@ SmallVector<int64_t> interchange(ArrayRef<int64_t> elements,
+
+// Creates a linalg.generic that transposes input using permutation indices.
+// Example: (M1, M0, N1, N0) -> (M1, N1, M0, N0) if indices = {0, 2, 1, 3}.
+Value createTransposeOp(Location loc, OpBuilder &builder, Value input,
```
note: a linalg.transpose up is coming upstream

MaheshRavishankar replied:
Is it more than a convenience method to create transpose?

chelini commented:
Looks good to me. Thanks for moving this forward.

nicolasvasilache commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/Vectorization.cpp:
```diff
@@ -0,0 +1,258 @@
+
+// Rewrites iree_linalg_ext.pack to tensor.pad + tensor.expand_shape +
+// linalg.generic (transpose) ops.
+struct GeneralizePackOpPattern : OpRewritePattern<PackOp> {
```
IMO we should avoid return failure if any of the outer dims is != 1.
I see you changed tests recently to all 1's.

Once we only accept leading 1's dimension, we should avoid reshapes and instead use rank-reducing insert/extract.

Then there would be no need for new vectorization patterns.

hanhanW commented:
Thanks for the review from all of you! After reading couple rounds of review comments, I'd like to clarify some things before updating the PR. Please correct me if I'm wrong.

There are two requirements based on the review comments:

1. outer dims should be all 1s during vectorization.
2. Having a new vectorization pattern is a red flag. We can make it work with generic vectorizer.

After thinking a while, here is the new vectorization process for pack op.

1. Tile the pack op and make the outer dims be all 1s. The iteration domain is over input domain, so the tile sizes would be
  a) `input[i]` if it's a untiled loop.
  b) `input[i] / mapped_inner_size[i]` if it's a tiled lop.

2. After tiling, the outer_dims are guaranteed to be all 1s. We can kick in the GeneralizePackOpPattern pattern.

3. Kick in generic vectorize to vectorize everything.

There was a problem about vector.shape_cast/tensor.expand_shape in step 2. My solution is to generate a special linalg transpose op in this case. Since the outer dims are all 1s, we can create a generic op like:

```
%trans = linalg.generic {
  indexing_maps = [affine_map<(d0, d1, d2, d3, d4, d5) -> (d5, d4)>,
                                 affine_map<(d0, d1, d2, d3, d4, d5) -> (d0, d1, d2, d3, d4, d5)>],
  iterator_types = ["parallel", "parallel", "parallel", "parallel", "parallel", "parallel"]}
  ins(%input : tensor<1x1x32x8xf32>)
  outs(%empty: tensor<1x1x1x1x8x32xf32>) {
   // ... yield directly
}
```

Or make the indexing_maps be `[affine_map<(d0, d1) -> (d1, d0)>, affine_map<(d0, d1) -> (0, 0, 0, 0, d0, d1)>]`. I assume that the generic vecotrizer can vectorize the transpose op and it avoids tensor.expand_shape ops.

Does the solution address all the concerns?

dcaballe commented:
I'm worried about the idea of having a separate vectorizer for these ops on the side of the main Linalg vectorizer. This is not going to compose well with things like masking and the rest of the vectorization features we are working on. Do you have an integration plan for this? I wonder if it would be easier to make the vectorization hooks of the main vectorizer available through the public API so that we can add external vectorization patterns. Would that work?

hanhanW commented:
> I'm worried about the idea of having a separate vectorizer for these ops on the side of the main Linalg vectorizer. This is not going to compose well with things like masking and the rest of the vectorization features we are working on. Do you have an integration plan for this? I wonder if it would be easier to make the vectorization hooks of the main vectorizer available through the public API so that we can add external vectorization patterns. Would that work?

Thanks for raising the concern! They are in a very early stage, and we do plan to upstream them after prototyping it in IREE. I'm not going to introduce new vectorization patterns or create a separate vectorizer in this PR. It's providing a way to vectorize pack op using existing generic vectorizer. The core idea is to rewrite the op into a generic op, and kick in the generic vectorizer, see https://github.com/iree-org/iree/pull/10734#issuecomment-1279314542 for more details. We're not going to introduce fancy vectorization patterns or strategy, so I'm pretty confident that it would compose well with the vectorization features we are working on.

hanhanW commented:
I remove vectorization patterns. The PR introduces generalization patterns for pack op, and creates a pass for vectorizing it. The purpose of the pass is mainly for testing.

The support for tiling outer dims will be done in other PRs. Because that requires more implementation (e.g., infer result shape when tiling with tile_size=inner_tile_size) which would make the PR big. I'd like to add the features incrementally.

hanhanW commented on llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/Passes/Passes.td:
```diff
@@ -59,6 +59,17 @@ def TopkSplitReduction:
 }
 
+def LinalgExtPackOpVectorization:
+    Pass<"iree-linalg-ext-pack-op-vectorization", "func::FuncOp"> {
```
I call it LinalgExtPackOpVectorization because the vectorization of unpack op is not clear to me. We could merge or rename the passes once we have clearer picture.

MaheshRavishankar replied:
Havent looked at the PR yet (probably later tonight or tomorrow), but I'd assume unpack vectorization would go similar. Unpack is transpose + collapse shape + extract_slice. With all the outer tile dimensions being 1, there would be no collapse shape. So just a transpose, and that get vectorized as other transposes.

hanhanW replied:
Good point! Yes, that's plan for vectorizing unpack ops. SG!

hanhanW commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Transforms.cpp:
```diff
@@ -784,6 +792,148 @@ createLinalgStrategyRemoveMarkersPass() {
+        AffineMap::get(nloops, 0, inputExprs, rewriter.getContext()),
+        AffineMap::getMultiDimIdentityMap(nloops, rewriter.getContext())};
+
+    auto transposedOp = rewriter.create<linalg::GenericOp>(
```
We're not able to use linalg.transpose because it requires the rank to be the same. Here we need a transpose op that expanding the shape (with unit dims).

MaheshRavishankar commented on llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Transforms.cpp:
```diff
@@ -784,6 +792,148 @@ createLinalgStrategyRemoveMarkersPass() {
+
+namespace {
+template <typename T>
+SmallVector<T> interchange(ArrayRef<T> elements,
```
Seems like we have this many places... might be worth moving to a utility function either in `LinalgExt` or in MLIR

hanhanW replied:
Yes, I'll do the cleanup afterwards.

MaheshRavishankar reviewed (approved):
Nice! Thanks! You can land this, but Nicolas might have suggestions that can be incorporated after the fact.

hanhanW commented:
Thanks for the review!! I think I addressed the concerns from Nicolas side. I'm going to land the PR. Nicolas, feel free to drop any comments. I'm happy to address them in post-fixes!
